### PR TITLE
[WIP] Add backpressure to pipelines

### PR DESCRIPTION
**Pull request by pakrym**
Still not sure about Flush returning bool

@davidfowl @halter73

### CODE REVIEW

File: src/System.IO.Pipelines/PipelineFactory.cs [hunk 1]
```diff
@@ -17,6 +17,7 @@ namespace System.IO.Pipelines
     public class PipelineFactory : IDisposable
     {
         private readonly IBufferPool _pool;
+        public int MaximumPipeSize { get; set; }
```
Comment (davidfowl) [inline src/System.IO.Pipelines/PipelineFactory.cs]: Should be an argument to create

**Comment by pakrym** [comment]
@davidfowl added min/max, internal

File: src/System.IO.Pipelines/Pipe.cs [hunk 2]
```diff
@@ -52,15 +53,23 @@ public class Pipe : IPipelineReader, IPipelineWriter, IReadableBufferAwaiter
...
         /// <summary>
         /// Initializes the <see cref="Pipe"/> with the specifed <see cref="IBufferPool"/>.
         /// </summary>
         /// <param name="pool"></param>
-        public Pipe(IBufferPool pool)
+        /// <param name="maximumSize"></param>
+        public Pipe(IBufferPool pool, long maximumSizeLow = 0, long maximumSizeHigh = 0)
         {
             _pool = pool;
+            _maximumSizeHight = maximumSizeHigh;
```
Comment (davidfowl) [inline src/System.IO.Pipelines/Pipe.cs]: _maximumSizeHight -> _maximumSizeHigh

File: src/System.IO.Pipelines/Pipe.cs [hunk 3]
```diff
@@ -393,6 +432,9 @@ public void AdvanceReader(ReadCursor consumed, ReadCursor examined)
 
             if (!consumed.IsDefault)
             {
+                var consumedBytes = ReadCursor.GetLength(_readHead, _readHead.Start, consumed.Segment, consumed.Index);
+                Interlocked.Add(ref _length, -consumedBytes);
```
Comment (davidfowl) [inline src/System.IO.Pipelines/Pipe.cs]: The locking around length is inconsistent. In some cases you update it inside of _sync and in this case you're using an interlocked. Move this code lower into the lock.

File: src/System.IO.Pipelines/Pipe.cs [hunk 4]
```diff
@@ -436,55 +475,32 @@ public void AdvanceReader(ReadCursor consumed, ReadCursor examined)
...
-        void IPipelineWriter.Complete(Exception exception) => CompleteWriter(exception);
 
-        /// <summary>
-        /// Marks the pipeline as being complete, meaning no more items will be written to it.
-        /// </summary>
-        /// <param name="exception">Optional Exception indicating a failure that's causing the pipeline to complete.</param>
-        public void CompleteWriter(Exception exception = null)
-        {
-            if (_producingState != State.NotActive)
+            if (_length < _maximumSizeLow)
```
Comment (davidfowl) [inline src/System.IO.Pipelines/Pipe.cs]: Here as well.

File: src/System.IO.Pipelines/FlushAsyncAwaitable.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime.CompilerServices;
+
+namespace System.IO.Pipelines
+{
+    public struct FlushAsyncAwaitable : ICriticalNotifyCompletion
```
Comment (KrzysztofCwalina) [inline src/System.IO.Pipelines/FlushAsyncAwaitable.cs]: Some of these types are what I would call "plumbing". We should consider moving them to some obscure sub namespace. But let's chat about it when we do an API review.

Comment (pakrym) [inline src/System.IO.Pipelines/FlushAsyncAwaitable.cs]: I'll make it internal for now.

File: tests/System.IO.Pipelines.Tests/BackpressureTests.cs [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+            var result = _pipe.ReadAsync().GetAwaiter().GetResult();
+            var consumed = result.Buffer.Move(result.Buffer.Start, 33);
+            _pipe.AdvanceReader(consumed, consumed);
+
+            Assert.True(flushAsync.IsCompleted);
+            Assert.True(flushAsync.Result);
+        }
+
+        [Fact]
+        public void FlushAsyncAwaitableDoesNotCompletesWhenReaderAdvancesUnderHight()
```
Comment (davidfowl) [inline tests/System.IO.Pipelines.Tests/BackpressureTests.cs]: High*

**Pull request merged**